src/main/io/gps.c

```diff
@@ -449,197 +572,132 @@ void gpsInitUblox(void)
+                        } else {
+                            ubloxSetMessageRate(CLASS_NAV, MSG_VELNED, 1); // set VELNED MSG rate
                         }
                         break;
-                    case UBLOX_ACK_WAITING:
-                        if ((++gpsData.ackTimeoutCounter) == UBLOX_ACK_TIMEOUT_MAX_COUNT) {
-                            gpsData.ackState = UBLOX_ACK_GOT_TIMEOUT;
+                    case 11:
+                        if (gpsData.ubloxUseSAT) {
+                            ubloxSetMessageRate(CLASS_NAV, MSG_SAT, 5); // set SAT MSG rate (every 5 cycles)
+                        } else {
+                            ubloxSetMessageRate(CLASS_NAV, MSG_SVINFO, 5); // set SVINFO MSG rate (every 5 cycles)
                         }
                         break;
-                    case UBLOX_ACK_GOT_TIMEOUT:
-                    case UBLOX_ACK_GOT_NACK:
-                    case UBLOX_ACK_GOT_ACK:
-                        gpsData.state_position = 0;
-                        gpsData.ackState = UBLOX_ACK_IDLE;
-                        gpsData.messageState++;
+                    case 12:
+                        ubloxSetNavRate(0xC8, 1, 1); // set rate to 5Hz (measurement period: 200ms, navigation rate: 1 cycle)
+                        break;
+                    case 13:
+                        ubloxSetSbas();
+                        break;
+                    case 14:
+                        if ((gpsConfig()->sbasMode == SBAS_NONE) || (gpsConfig()->gps_ublox_use_galileo)) {
+                            ubloxSendPollMessage(MSG_CFG_GNSS);
+                        } else {
+                            gpsSetState(GPS_RECEIVING_DATA);
+                        }
                         break;
                     default:
                         break;
                 }
             }
-
-            if (gpsData.messageState >= GPS_MESSAGE_STATE_INITIALIZED) {
-                // ublox should be initialised, try receiving
-                gpsSetState(GPS_RECEIVING_DATA);
+        
+            switch (gpsData.ackState) {
+                case UBLOX_ACK_IDLE:
+                    break;
+                case UBLOX_ACK_WAITING:
+                    if ((++gpsData.ackTimeoutCounter) == UBLOX_ACK_TIMEOUT_MAX_COUNT) {
+                        gpsSetState(GPS_LOST_COMMUNICATION);
+                    }
+                    break;
+                case UBLOX_ACK_GOT_ACK:
+                    if (gpsData.state_position == 14) {
+                        // ublox should be initialised, try receiving
+                        gpsSetState(GPS_RECEIVING_DATA);
+                    } else {
+                        gpsData.state_position++;
+                        gpsData.ackState = UBLOX_ACK_IDLE;
+                    }
+                    break;
+                case UBLOX_ACK_GOT_NACK:
+                    if (gpsData.state_position == 7) { // If we were asking for NAV-PVT...
```

Comment:
`gpsData.state_position == 7` can cause serios headaches especially with `gpsData.state_position++` in the previous part. I haven't got better idea yet.

Reply:
Indeed, without co-routines (or similar) support I don't have a better idea without overcomplicating this particular case